pkg/registry/service/allocator/allocator.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocator
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"net"
+	"sync"
+)
+
+// Interface manages the allocation of IP addresses out of a range. Interface
+// should be threadsafe.
+type Interface interface {
+	Allocate(net.IP) error
+	AllocateNext() (net.IP, error)
+	Release(net.IP) error
+}
+
+// Snapshottable is an Interface that can be snapshotted and restored. Snapshottable
+// should be threadsafe.
+type Snapshottable interface {
+	Interface
+	Snapshot() (*net.IPNet, []byte)
+	Restore(*net.IPNet, []byte) error
+}
+
+var (
+	ErrFull               = errors.New("range is full")
+	ErrNotInRange         = errors.New("provided IP is not in the valid range")
+	ErrAllocated          = errors.New("provided IP is already allocated")
+	ErrMismatchedNetwork  = errors.New("the provided network does not match the current range")
+	ErrAllocationDisabled = errors.New("IP addresses cannot be allocated at this time")
+)
+
+// Range is a contiguous block of IPs that can be allocated atomically.
+type Range struct {
+	net *net.IPNet
+	// base is a cached version of the start IP in the CIDR range as a *big.Int
+	base *big.Int
+	// strategy is the strategy for choosing the next available IP out of the range
+	strategy allocateStrategy
+	// max is the maximum size of the range
+	max int
+
+	// lock guards the following members
+	lock sync.Mutex
+	// count is the number of currently allocated elements in the range
+	count int
+	// allocated is a bit array of the allocated ips in the range
+	allocated *big.Int
+}
+
+// allocateStrategy is a search strategy in the allocation map for a valid IP.
+type allocateStrategy func(allocated *big.Int, max, count int) (int, error)
+
+// NewCIDRRange creates a Range over a net.IPNet.
+func NewCIDRRange(cidr *net.IPNet) *Range {
+	ones, bits := cidr.Mask.Size()
+	max := int(1) << uint(bits-ones)
+	r := Range{
+		net:      cidr,
+		strategy: randomScanStrategy,
+		base:     bigForIP(cidr.IP),
+		max:      max,
+
+		allocated: big.NewInt(0),
+		count:     0,
+	}
+	return &r
+}
+
+// Free returns the count of IP addresses left in the range.
+func (r *Range) Free() int {
```

Comment:
Could be, but I wanted an allocator to be able to check remaining eventually to handle heuristics and warnings.  It's not on the allocator.Interface, but it's something the etcd service controller might want to print (and should probably include max as well)

----- Original Message -----
> > +	ones, bits := cidr.Mask.Size()
> > +	max := int(1) << uint(bits-ones)
> > +	r := Range{
> > +		net:      cidr,
> > +		strategy: randomScanStrategy,
> > +		base:     bigForIP(cidr.IP),
> > +		max:      max,
> > +
> > +		allocated: big.NewInt(0),
> > +		count:     0,
> > +	}
> > +	return &r
> > +}
> > +
> > +// Free returns the count of IP addresses left in the range.
> > +func (r *Range) Free() int {
> 
> afaict this could be an unexported symbol?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29878931